### ISSUE TITLE
Refreshing the file without closing and reopening its respective widget?

### ISSUE THREAD

**Issue by ashleyswang** [jupyterlab/jupyterlab#8637]
I am trying to make a JupyterLab extension that will pull from a Git repo and show the changes on JupyterLab. I currently have all the Git functionality working, but I'm having issues trying to make the changes show for files that are already opened. For example if I have `hi.txt` opened in JupyterLab, but I receive changes from a Git pull, the changes will not show on the file editor unless I close it and reopen it. 

I have tried playing around with the `DocumentManager` class (specifically the `open()` function), but it seems that the same context for the `DocumentWidget` is used. So even if I am able to open up new tabs of the same file, the same context for the already opened file is used, and therefore no changes are being shown in the new tabs. This means that I need to make sure the existing context is disposed before I can reopen a new context. 

Since I would like be able to use the same widget (so from the user perspective the changes show up on the same file editor without some weird flashing of the file editor closing and reopening), I thought that replacing the context might work, but since most useful functions relating to contexts are private (including `_createContext`), I don't know if this is a viable option anymore.

TLDR: Does anyone know of a good way to refresh the contents in the file editor without having to close and reopen the entire widget?

**Comment by jasongrout** [comment]
The "File > Reload ... from disk" menu item is exactly for the case where the file content changed on disk and we want to pull the updated content into JupyterLab.

**Comment by ashleyswang** [comment]
Thank you! That seemed to work, but I was wondering if there was a way to access the contents within the file editor to read them. Ideally I'd be able to load changes while still maintaining the cursor position/changes made by the user.

**Comment by jasongrout** [comment]
Perhaps you want to fetch the file contents yourself, using the services api, then do the diff with what is in the current document model?

**Comment by ashleyswang** [comment]
Okay thank you! I'll try that. Do you also happen to know if there's any good way to determine all the files that are currently opened on widgets? I am trying to get something similar to the document manager's `_contexts` private member.
